### PR TITLE
FUSETOOLS2-1822: UI test for value update with debugger

### PR DESCRIPTION
**Pull request by unsortedhashsets**
Related: https://issues.redhat.com/browse/FUSETOOLS2-1822

Test result:

Screenshot from 2023-06-05 13-30-42

### CODE REVIEW

**Comment by djelinek** [comment]
test looks fine, but it would be nice to handle SonarCloud code analysis report

**Review by mlorinc — CHANGES_REQUESTED** [review]

File: src/ui-test/tests/debugger.test.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        expect(result).to.be.true;
+
+        // WORKAROUND: https://github.com/redhat-developer/vscode-extension-tester/issues/402
+        await driver.wait(async function () {
+            try {
+                const debugView = (await (await new ActivityBar().getViewControl('Run')).openView()) as DebugView;
+                const variables = await debugView.getContent().getSection('Variables');
+                const messages = await variables.openItem('Message');
+                for await (let message of messages) {
+                    if (await message.getAttribute('aria-label') === `Body, value ${HELLO_CAMEL_MESSAGE}`) {
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:74]: Will take a look, but for now I see that element very not deterministic to make some generic solution

File: src/ui-test/tests/debugger.test.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    after('After cleanup', async function () {
+        await new EditorView().closeAllEditors();
+    });
+
+    beforeEach('Before each test', async function () {
+        await VSBrowser.instance.openResources(path.resolve('src', 'ui-test', 'resources'));
+        await VSBrowser.instance.waitForWorkbench();
+
+        await (await new ActivityBar().getViewControl('Explorer') as ViewControl).openView();
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:44]: Not needed, thanks! nice catch

File: src/ui-test/tests/debugger.test.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+        const editorView = new EditorView();
+        await driver.wait(async function () {
+            return (await editorView.getOpenEditorTitles()).find(title => title === CAMEL_ROUTE_YAML_WITH_SPACE);
+        }, 5000);
+    });
+
+    afterEach('After each test', async function () {
+        await disconnectDebugger(driver);
+        await (await new ActivityBar().getViewControl('Run and Debug') as ViewControl).closeView();
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:57]: Not needed, thanks! nice catch

File: src/ui-test/tests/debugger.test.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        driver = VSBrowser.instance.driver;
+    });
+
+    after('After cleanup', async function () {
+        await new EditorView().closeAllEditors();
+    });
+
+    beforeEach('Before each test', async function () {
+        await VSBrowser.instance.openResources(path.resolve('src', 'ui-test', 'resources'));
+        await VSBrowser.instance.waitForWorkbench();
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:42]: @djelinek Hi, you also use it in tests :thinking: , I will test that test without `await VSBrowser.instance.waitForWorkbench();`

File: src/ui-test/tests/debugger.test.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        await section.openItem(CAMEL_ROUTE_YAML_WITH_SPACE);
+
+        const editorView = new EditorView();
+        await driver.wait(async function () {
+            return (await editorView.getOpenEditorTitles()).find(title => title === CAMEL_ROUTE_YAML_WITH_SPACE);
+        }, 5000);
+    });
+
+    afterEach('After each test', async function () {
+        await disconnectDebugger(driver);
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:56]: Agreed :thinking:

File: src/ui-test/tests/debugger.test.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        await (await new ActivityBar().getViewControl('Run and Debug') as ViewControl).closeView();
+        await killTerminal();
+    });
+
+    it('Update of a value with Camel debugger', async function () {
+        await executeCommand(CAMEL_RUN_DEBUG_ACTION_LABEL);
+        await waitUntilTerminalHasText(driver, TEST_ARRAY_RUN_DEBUG);
+        let result = await new TextEditor().toggleBreakpoint(11);
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:64]: Agreed :thinking:

File: src/ui-test/tests/debugger.test.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    if (await message.getAttribute('aria-label') === `Body, value ${HELLO_CAMEL_MESSAGE}`) {
+                        await driver.actions().doubleClick(message).perform();
+                        await driver.actions().clear();
+                        await driver.actions().sendKeys(HELLO_WORLD_MESSAGE).perform();
+                        return true;
+                    }
+                }
+                return false;
+            } catch (e) {
+                await driver.actions().click().perform();
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:83]: During tests I have faced here two types of errors:
1) Object is not ready
2) And popup that blocks any next action (on pic.)

Screenshot from 2023-06-06 10-46-41


previously we faced issues with popups -> https://issues.redhat.com/browse/FUSETOOLS2-2100

and solution was just to make extra click in catch section to close the popup

[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:83]: mainly we face that issue so much due the fact that on GitHub actions machines we have very small resolution

[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:83]: Will add a comment

File: src/ui-test/tests/debugger.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlorinc) [reply src/ui-test/tests/debugger.test.ts:71]: I will try one solution as well, as I have got an idea yesterday while falling asleep.

File: src/ui-test/tests/debugger.test.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        await disconnectDebugger(driver);
+        await (await new ActivityBar().getViewControl('Run and Debug')).closeView();
+        await killTerminal();
+        await new EditorView().closeAllEditors();
+    });
+
+    it('Update of a value with Camel debugger', async function () {
+        await executeCommand(CAMEL_RUN_DEBUG_ACTION_LABEL);
+        await waitUntilTerminalHasText(driver, TEST_ARRAY_RUN_DEBUG);
+        await driver.wait(async function () {
```
Comment (mlorinc) [inline src/ui-test/tests/debugger.test.ts:60]: Please verify timeout is finite in that case. I could not find it in Selenium documentation.

**Review by mlorinc — APPROVED** [review]

File: src/ui-test/tests/debugger.test.ts (hunk 8 above)
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:60]: :thinking: veeery nice catch, tested it is >100s, fixed

File: src/ui-test/tests/debugger.test.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        await section.openItem(CAMEL_ROUTE_YAML_WITH_SPACE);
+
+        const editorView = new EditorView();
+        await driver.wait(async function () {
+            return (await editorView.getOpenEditorTitles()).find(title => title === CAMEL_ROUTE_YAML_WITH_SPACE);
+        }, 5000);
+    });
+
+    after('After cleanup', async function () {
+        await driver.wait(async function () {
```
[the remark this answers is not in the document]
Comment (unsortedhashsets) [reply src/ui-test/tests/debugger.test.ts:48]: :thinking: veeery nice catch, tested it is >100s, fixed

**Review by djelinek — APPROVED** [review]

**Review by mlorinc — APPROVED** [review]

**Pull request merged**